parser/src/test/java/com/facebook/coresql/parser/sqllogictest/java/SqlLogicTest.java

```diff
@@ -0,0 +1,272 @@
+        writer.write(format("%s%% of sqllogictest statements were successfully parsed.",
+                new DecimalFormat("#.##").format(numStatementsParsed / (double) numStatementsFound * 100)));
+        writer.close();
+    }
+
+    private void validateParsingOfSqlStatement(String statement)
+            throws IOException
+    {
+        String errorType = "";
+        try {
+            errorType = SqlLogicTestError.PARSING_ERROR.toString();
+            AstNode ast = parseStatement(statement);
+            if (ast == null) {
+                throw new Exception();
+            }
+            errorType = SqlLogicTestError.UNPARSING_ERROR.toString();
+            String unparsed = unparse(ast);
+            errorType = SqlLogicTestError.UNPARSED_DOES_NOT_MATCH_ERROR.toString();
+            if (!statement.trim().equals(unparsed.trim())) {
+                throw new Exception();
+            }
+            errorType = "";
+        }
+        catch (Exception ignored) {
+        }
+
+        if (errorType.equals("")) {
+            numStatementsParsed += 1;
+        }
+        else {
+            writer.write(statement + " ----> " + errorType);
+            writer.write("\n\n");
+        }
+    }
+
+    private class SqlLogicTestFileVisitor
+            implements FileVisitor<Path>
+    {
+        private void processStatement(BufferedReader reader)
+                throws IOException
+        {
+            StringBuilder statement = new StringBuilder();
+            while (true) {
+                final String line = reader.readLine();
+                if (line == null || isLineAfterStatement(line)) {
+                    break;
+                }
+                statement.append(line);
+            }
+            if (!statement.substring(statement.length() - 1).equals(";")) {
+                statement.append(";"); // NOTE: This semi-colon is necessary for parsing.
+            }
+            numStatementsFound += 1;
+            validateParsingOfSqlStatement(statement.toString());
+        }
+
+        private boolean isLineBeforeStatement(String line)
+        {
+            return line.startsWith("statement ok") || line.startsWith("query");
+        }
+
+        private boolean statementIsAcceptableToCurrRdms()
+        {
+            if (rdms == AcceptedRdms.ALL) {
+                return skipIfSet.isEmpty() && onlyIf == null;
+            }
+
+            if (onlyIf != null && onlyIf != rdms) {
+                return false;
+            }
+            return !skipIfSet.contains(rdms);
+        }
+
+        private boolean isLineAfterStatement(String line)
+        {
+            return line.equals("----") || line.equals("");
+        }
+
+        private boolean processConditionalRecord(String line)
+        {
+            if (line.startsWith("onlyif")) {
+                String onlyIfRdms = line.split(" ")[1];
```

Comment:
Could this throw an IndexOutOfBoundsException?

Reply:
Potentially -- refactored to ensure this exception is explicitly handled.

Done.